pre-commit.sh

```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
+
+# shellcheck disable=SC2034
+# Load git diff once for slight performance improvements
+git_changes=$(git diff --name-only)
+
+execute_or_run() {
+    folder="$1"
+    container_name="$2"
+    service="$3"
+    cmd="$4"
+    use_cmd_as_entrypoint="${5:-false}" # default to false if not provided
+
+    if echo "$git_changes" | grep -qE "^$folder/"; then
+        if docker inspect --format="{{.State.Running}}" "$container_name" 2>/dev/null | grep -q "true"; then
+            docker compose exec -d "$service" $cmd
+        else
+            # Check if we should use the command as the entrypoint
+            if $use_cmd_as_entrypoint; then
+                docker compose run --rm -d --entrypoint="$cmd" "$service"
+            else
+                docker compose run -d "$service" $cmd
+            fi
+        fi
+    fi
+}
+
+
+# Frontend
+execute_or_run "frontend" "ecamp3-frontend" "frontend" "npm run lint" &
```

Comment:
do you put the processes in the background to speed this up?
I would argue that its simpler if the steps are run sequentially and that they are fast enough.
(They are fast enough if you haven't checked out the repo in the windows file system.
File system access from the WSL2 docker desktop backend to the windows file system get slow in my experience, especially if you have many files, like in the node_modules folder)

Reply:
> do you put the processes in the background to speed this up?

Yes. When I tested it, it was faster (my tests were very superficial and just one system). I believe that for most commits there are changes in one or two of the checked folders, if nothing is changed it will not be linted. 

> I would argue that its simpler if the steps are run sequentially and that they are fast enough.

I had the steps run sequentially first and then tried to make speed optimisations since this is a pre-commit hook. I noticed that the relative speed difference is very notable if there are no changes in a folder (which is the most likely case). I don't understand much about Unix parallel/sequential runs so it would be wrong to insist on keeping it. However, I don't think I understand your reasoning for sequential runs except for the 'its simpler'-part. Would you mind elaborating?

Reply:
To do this really right you maybe need to add traps to forward the stop or kill signals to the child processes.
And all the other things i forgot.
i rarely need to run the linter on more than one directory.
For me running the commands sequentially and for the one project i need it is enough.
but it is ok if you want to keep it